packages/jsii-pacmak/lib/targets/java.ts

```diff
@@ -1513,12 +1522,211 @@ class JavaGenerator extends Generator {
+      this.code.openBlock(
+        `for (int ${varName} = 0; ${varName} < ${value}.size(); ${varName}++)`,
+      );
+      validate.call(
+        this,
+        `${value}.get(${varName})`,
+        `${descr}.append(".get(").append(${varName}).append(")")`,
+        elementType,
+        parameterName,
+      );
+      this.code.closeBlock();
+    }
+
+    function validateMap(
+      this: JavaGenerator,
+      value: string,
+      descr: string,
+      elementType: spec.TypeReference,
+      parameterName: string,
+    ) {
+      const varName = `__item_${createHash('sha256')
+        .update(descr)
+        .digest('hex')
+        .slice(0, 6)}`;
+      const javaElemType = this.toJavaType(elementType);
+      this.code.openBlock(
+        `for (java.util.Map.Entry<String, ${javaElemType}> ${varName}: ${value}.entrySet())`,
+      );
+      validate.call(
+        this,
+        `${varName}.getValue()`,
+        `${descr}.append(".get(\\"").append((${varName}.getKey())).append("\\")")`,
+        elementType,
+        parameterName,
+      );
+      this.code.closeBlock();
+    }
+
+    function validateTypeUnion(
+      this: JavaGenerator,
+      value: string,
+      descr: string,
+      type: spec.UnionTypeReference,
+      parameterName: string,
+    ) {
+      this.code.indent('if (');
+      let emitAnd = false;
+      const nestedCollectionUnionTypes = [];
+      const typeRefs = type.union.types;
+      for (const typeRef of typeRefs) {
+        const prefix = emitAnd ? '&&' : '';
+        const javaType = this.toJavaTypeNoGenerics(typeRef);
+        if (javaType !== this.toJavaType(typeRef)) {
+          nestedCollectionUnionTypes.push(typeRef);
+        }
+        const test = `${value} instanceof ${javaType}`;
+        this.code.line(`${prefix} !(${test})`);
+        emitAnd = true;
+      }
+      // Only anonymous objects at runtime can be `JsiiObject`s.
+      this.code.line(
+        `&& !(${value}.getClass().equals(software.amazon.jsii.JsiiObject.class))`,
+      );
+
+      this.code.unindent(')');
+      this.code.openBlock('');
+
+      const placeholders = typeRefs
+        .map((typeRef) => {
+          return `${this.toJavaType(typeRef)}`;
+        })
+        .join(', ');
+
+      this.code.line(`throw new IllegalArgumentException(`);
+      this.code.line(`new java.lang.StringBuilder("Expected ")`);
+      this.code.line(`${descr}`);
+      this.code.line(`.append(" to be one of: ")`);
+      this.code.line(`.append("${placeholders}")`);
+      this.code.line(`.append("; received ")`);
+      this.code.line(`.append(${value}.getClass())`);
+      this.code.line(
+        `.append(", (Parameter '${parameterName}')").toString());`,
+      );
+      this.code.closeBlock();
+
+      for (const typeRef of nestedCollectionUnionTypes) {
+        this.code.openBlock(
+          `if (${value} instanceof ${this.toJavaTypeNoGenerics(typeRef)})`,
+        );
+        validate.call(this, value, descr, typeRef, parameterName);
```

Comment:
This leads to invalid Java apparently

```suggestion
        const varName = `__cast_${createHash('sha256').update(value).digest('hex').slice(0, 6)}`;
        const javaTypeFull = this.toJavaType(typeRef);
        this.code.line(`@SuppressWarnings("unchecked")`);
        this.code.line(`final ${javaTypeFull} ${varName} = (${javaTypeFull})${value};`);
        validate.call(this, varName, descr, typeRef, parameterName);
```

Reply:
could you elaborate? I'm not able to reproduce this.